internal/logger/logger.go

```diff
@@ -0,0 +1,249 @@
+		Level:                 "info",
+	})
+}
+
+// InitLogger sets up the logging framework
+//
+// In production, the container logs will be collected and file logging should be disabled. However,
+// during development it's nicer to see logs as text and optionally write to a file when debugging
+// problems in the containerized pipeline
+//
+// The output log file will be located at /var/log/service-xyz/service-xyz.log and
+// will be rolled according to configuration set.
+// To disable usage for unit tests, just pass in the Config propery Disable: true.
+func InitLogger(config Config) *Logger {
+	var writers []io.Writer
+
+	if !config.Enable {
+		zerolog.SetGlobalLevel(zerolog.Disabled)
+	} else {
+		// Setting as switch so I can expand later to other layers like trace etc if required
+		switch config.Level {
+		case "info":
+			zerolog.SetGlobalLevel(zerolog.InfoLevel)
+		case "debug":
+			zerolog.SetGlobalLevel(zerolog.DebugLevel)
+		default:
+			zerolog.SetGlobalLevel(zerolog.InfoLevel)
+		}
+	}
+
+	consoleWriter := zerolog.ConsoleWriter{Out: os.Stdout, TimeFormat: time.RFC3339}
+
+	if config.ConsoleLoggingEnabled {
+		consoleWriter.FormatLevel = func(i interface{}) string {
+			return strings.ToUpper(fmt.Sprintf("| %-6s|", i))
+		}
+		consoleWriter.FormatMessage = func(i interface{}) string {
+			return fmt.Sprintf("%s |", i)
+		}
+		consoleWriter.FormatFieldName = func(i interface{}) string {
+			return fmt.Sprintf("%s:", i)
+		}
+		consoleWriter.FormatFieldValue = func(i interface{}) string {
+			return strings.ToUpper(fmt.Sprintf("%s", i))
+		}
+		writers = append(writers, consoleWriter)
+	}
+
+	if config.FileLoggingEnabled {
+		writers = append(writers, newRollingFile(config))
+	}
+
+	// mw := io.MultiWriter(writers...)
+	multi := zerolog.MultiLevelWriter(writers...)
+
+	// zerolog.SetGlobalLevel(zerolog.DebugLevel)
+	// logger := zerolog.New(mw).With().Timestamp().Logger()
+	logger := zerolog.New(multi).With().Timestamp().Caller().Logger()
+
+	logger.Info().
+		Bool("fileLogging", config.FileLoggingEnabled).
+		Bool("jsonLogOutput", config.EncodeLogsAsJson).
+		Str("logDirectory", config.Directory).
+		Str("fileName", config.Filename).
+		Int("maxSizeMB", config.MaxSize).
+		Int("maxBackups", config.MaxBackups).
+		Int("maxAgeInDays", config.MaxAge).
+		Str("level", config.Level).
+		Msg("logging configured")
+
+	return &Logger{&logger}
+}
+
+func newRollingFile(config Config) io.Writer {
+	if err := os.MkdirAll(config.Directory, 0o755); err != nil {
+		log.Error().Err(err).Str("path", config.Directory).Msg("can't create log directory")
+
+		return nil
+	}
+
+	return &lumberjack.Logger{
+		Filename:   path.Join(config.Directory, config.Filename),
+		MaxBackups: config.MaxBackups, // files
+		MaxSize:    config.MaxSize,    // megabytes
+		MaxAge:     config.MaxAge,     // days
+	}
+}
+
+// // Output duplicates the global logger and sets w as its output.
+// func (l *Logger) Output(w io.Writer) zerolog.Logger {
```

Comment:
This is what I wanted to avoid having to do. I want to embed the pointer to `*zerolog.Logger` so that all the original methods of zerolog are available, and I've only provided the intialization and setup logic in a consistent way (along with file rotation, additional context fields, etc).